torchsde/_core/base_sde.py

```diff
@@ -114,10 +70,120 @@ def gdg_prod(self, t, y, v):
         pass
 
 
-class SDEStratonovich(BaseSDE):
+class ForwardSDE(BaseSDE):
 
-    def __init__(self, noise_type):
-        super(SDEStratonovich, self).__init__(noise_type=noise_type, sde_type='stratonovich')
+    def __init__(self, base_sde):
+        super(ForwardSDE, self).__init__(sde_type=base_sde.sde_type, noise_type=base_sde.noise_type)
+        self._base_sde = base_sde
+
+        # Register the core function. This avoids polluting the codebase with if-statements.
+        self.g_prod = {
+            settings.NOISE_TYPES.diagonal: self.g_prod_diagonal,
+            settings.NOISE_TYPES.additive: self.g_prod_additive,
+            settings.NOISE_TYPES.scalar: self.g_prod_scalar,
+            settings.NOISE_TYPES.general: self.g_prod_general
+        }[base_sde.noise_type]
+        self.gdg_prod = {
+            settings.NOISE_TYPES.diagonal: self.gdg_prod_diagonal,
+            settings.NOISE_TYPES.additive: self._skip,
+            settings.NOISE_TYPES.scalar: self.gdg_prod_scalar,
+            settings.NOISE_TYPES.general: self.gdg_prod_general
+        }
+        self.gdg_jvp = {
+            settings.NOISE_TYPES.diagonal: self._skip,
+            settings.NOISE_TYPES.additive: self._skip,
+            settings.NOISE_TYPES.scalar: self._skip,
+            settings.NOISE_TYPES.general: self.gdg_jvp_v2
+        }[base_sde.noise_type]
+
+    def f(self, t, y):
+        return self._base_sde.f(t, y)
+
+    def g(self, t, y):
+        return self._base_sde.g(t, y)
+
+    def h(self, t, y):
+        return self._base_sde.h(t, y)
+
+    # g_prod functions.
+    def g_prod_diagonal(self, t, y, v):
+        return misc.seq_mul(self._base_sde.g(t, y), v)
+
+    def g_prod_additive(self, t, y, v):
+        return self.g_prod_general(t, y, v)
+
+    def g_prod_scalar(self, t, y, v):
+        return misc.seq_mul_bc(self._base_sde.g(t, y), v)
+
+    def g_prod_general(self, t, y, v):
+        return misc.seq_batch_mvp(ms=self._base_sde.g(t, y), vs=v)
+
+    # gdg_prod functions.
+    def gdg_prod_general(self, t, y, v):
+        # TODO: Write this!
+        raise NotImplemented
+
```

Comment:
Thanks for catching this! In other places `gdg_prod` is used for Milstein, but I just realized for general noise, we need also supply the Levy area, so in some sense this function signature is wrong. 

Since we'll be using the jvp/contraction stuff, I'm putting a TODO here with some documentation about it.